documentation/smg/10_build_release_process.md

```diff
@@ -0,0 +1,89 @@
+
+<img src="./images/10_git_hotfix_branches.png">
+
+1. User reports issue
+2. Bug report issue created in GitHub/Issues documenting how to reproduce
+3. Fix developed on a branch (`xxx_branch`) taken from the *release* tag (`Rm_n_o`)
+4. All tests likely to be affected by changes in Herbert and Horace must be run on developers machine
+5. PR created for merge into Release branch (`rel_m_n_o`) *and* `master` branch. If this is the first release patch, the branch will need to be created
+6. If the build / test against of the PR against the Release branch succeeds on all platforms, that build artifact can be released to the target platform as version `m.n.o+1`
+
+If the hotfix is being done out-of-hours the PR can be merged by the developer WITHOUT review, but the branch should not be deleted and should be reviewed as the earliest opportunity.
+
+The same caveats apply about the Horace build dependencies on Herbert.
+
+Advantages:
+
+- build tools are not required on the production hardware
+- all builds are created using the same build tools
+- users are using a known, traceable, validated release
+
+Disadvantages:
+
+- slower; have to wait for full build and test pipeline to execute
+
+### Releases
+
+The Jenkins build artifacts are are not accessible outside of STFC. End users will be directed to GitHub to access packaged releases.
+
+A 'deploy' pipeline will be created ([Horace #73](https://github.com/pace-neutrons/Horace/issues/73)). 
+
+This will:
+
+- tag the `master` branch with the release number (`Rm_n_o`)
+- copy the built artifact to the GitHub
+
+The release tag will be an anchor for any subsequently needed release branch and patch branches.
+
+The version number is stored in a text file (`VERSION`) in the root of the Herbert and Horace source. This will follow [semantic versioning](https://semver.org/) and is used in the build process to set the version number in the MATLAB and compiled C++ components.
+
+### Facility Deployment
+
+Users participating in ISIS experiments use expensive neutron facility and need to be able to see and analyse their data as quickly as possible, ideally in the process of experiment and immediately after that. In addition to standard Horace scripts users sometimes write their own scripts, which places high demand on the software reliability. Sometimes, their custom scripts use standard software in the operational modes, which are difficult to predict. Often these script identify issues, which have not been anticipated during development so users need urgent fixing or modifications to the software. Sometimes, the issues can be caused by specific user's data, so can be easily reproduced only using these data. The changes fixing the user issue should occur immediately, as users should not lose a day or even an hour waiting until the whole build pipeline completes, produces all possible artefacts and get installed on the facility computers.  
+
```

Comment:
If we invest more time in developing tests around these scenarios it should lead to reducing these effects. We need to ensure we have a mechanism for capturing these issues (and possibly generalising them) or else we will forever be firefighting with hotfixes. We should make this a driver to improve quality control as opposed to a normal operating mode.

Reply:
```
Often these script identify issues, which have not been anticipated during development so users need urgent fixing or modifications to the software. Sometimes, the issues can be caused by specific user's data, so can be easily reproduced only using these data. 
```
This is exactly the use-case for the "Hotfix" workflow described elsewhere in the document. The user script and data should be referenced/included in the issue ticket.

We should aim to minimise these failures though (increasingly comprehensive) testing, but agree that there will always be cases we haven't seen.

If your immediate fixes are rolled out without comprehensive testing there is a real danger of unintended bugs being introduced and Horace breaking for all the other users on the machine (as there is a common/shared code base). 

A decision must be reached on what "level" of support the PACE tooling has in production. What is an acceptable turn around time for a bug fix? In what situations could a workaround be identified more rapidly requiring a only small change to the users custom script, for example.

Reply:
I am just worried we are indirectly creating a requirement to have a dev environment on every production instance.

Reply:
until I have not fixed an issue during for at least three cycles in a row, the results of the investing in the tests will not be conclusinve so workflow is necessary.  At the moment I am fixing an average 3 issues per cycle. The waste majority is minor.

Reply:
We are not creating the need for the enviroment -- Matlab is there anyway, and we are not chaning anything else. 

I am describing current level of the support. If somebody takes this task from me, he will decide on the best way of proceed. This way is proven to be successful for now. 

It also allows rapid introducing of the new features (equivalent of Mantid "nightly" edition)
As you know, this feature is in high demand on Mantid. Do not think we need it exactly Mantid way, as team and demands are order of magnitude smaller, but further morphing may come from this.

Reply:
Nightlies are good strategies. But changes in our nightlies go through the full PR/Review/Build/Package process. There are no hotfixes applied in situ. Users may have to wait at least 24 hours for the new nightly build. Which is fundamentally different from what you are suggesting.

Reply:
The "nightly" (aka 'master') build can be manually triggered so the wait time would only be the combined Horace and Herbert build/test time, ~2.5hrs at the moment.

Reply:
Another thought triggered while summarising this discussion is how this model maps onto the SCARF deployment when that becomes available to users.

If a user encounters issues on iDAaaS they are likely to hit the same issue when running the analysis on bigger data on the SCARF cluster

Reply:
looks like scarf deployment will me mainly compiled code deployment, so will not receive hotfixes.